auth/src/AzureSessionProvider.ts

```diff
@@ -0,0 +1,25 @@
+import { TenantIdDescription } from "@azure/arm-resources-subscriptions";
+import { AuthenticationSession, Event } from "vscode";
+
+export type SignInStatus = "Initializing" | "SigningIn" | "SignedIn" | "SignedOut";
+
+export type AzureAuthenticationSession = AuthenticationSession & {
+    tenantId: string;
+};
+
+export type DefinedTenant = TenantIdDescription & Required<Pick<TenantIdDescription, "tenantId" | "displayName">>;
+
+export enum GetSessionBehavior {
+    Silent,
+    PromptIfRequired,
+}
+
+export type AzureSessionProvider = {
+    signIn(): Promise<void>;
+    signInStatus: SignInStatus;
+    tenants: DefinedTenant[];
+    isSignedInToTenant(tenantId: string): boolean;
+    signInStatusChangeEvent: Event<SignInStatus>;
+    getAuthSession(tenantId: string, behavior: GetSessionBehavior, scopes?: string[]): Promise<AzureAuthenticationSession | undefined>;
```

Comment:
I think it's worth reconsidering the parameters for `getAuthSession`. For my purposes at least, the only reason for including `scopes` in there was to specify the special `VSCODE_CLIENT_ID` value.

The way I've done it here, it's a bit weird that the special `TENANT_ID` scope is abstracted away, but consumers are expected specify `VSCODE_CLIENT_ID` explicitly.

Could we consider something like an `options` object? E.g.
```typescript
export type GetAuthSessionParams {
    tenantId: string;
    behavior: GetSessionBehavior;
    applicationClientId?: string;
}
```

If there's a need to specify unconstrained `scopes` it could be added later. However, if we add it now and it causes confusion ("what scopes do I _actually_ need to specify?"), it'll be hard to remove.

Reply:
1. I'm not familiar with the special `VSCODE_CLIENT_ID`? How and why are you using this?
2. And I agree, I think an options object would be good here. We can add an optional scopes property in there too.

```ts
export type GetAuthSessionParams {
    tenantId: string;
    behavior: GetSessionBehavior;
    applicationClientId?: string;
    scopes?: string[];
}
```

Reply:
Also, there are scenarios where my team has needed to exclude the ARM scope, and add in something else. I will research into exactly what that was and report back. I want to make sure we can cover that scenario too.

Reply:
> I'm not familiar with the special VSCODE_CLIENT_ID? How and why are you using this?

I'm not _yet_, but I'm experimenting with using the session credentials to access `graph` endpoints for listing and creating applications and service principals. The [default "Visual Studio Code" application](https://github.com/microsoft/vscode/blob/4b81db67db61a03b3cf795a37ee8a7aaa4924f94/extensions/microsoft-authentication/src/AADHelper.ts#L21) does not have the required delegated permissions to do that. You can use the `VSCODE_CLIENT_ID` scope to [specify a different application](https://www.eliostruyf.com/microsoft-authentication-provider-visual-studio-code/#using-the-authentication-provider-with-your-azure-ad-app).

Speaking of that, when I said above I wasn't using the `scopes` option I think I was wrong. When creating an authentication provider needed to initialise the graph client, you need to implement a `getAccessToken` function. This takes some optional `scopes` and I'm passing them on to my `getAuthSession` function. If I didn't do that, I think the `aud` claim in the resulting access token would be wrong, and requests would be rejected by the graph endpoints.

Perhaps the scenarios you're referring to above are similar (e.g. accessing graph rather than ARM).